> Btw, I think we could add a generic `protoutil.UnmarshalNew()` which always creates a new message, guaranteeing that it's safe to omit `Reset()`. Something like:
> 
> ```go
> func UnmarshalNew[M Message](data []byte) (M, error) {
>   var m M
>   return m.Unmarshal(data)
> }
> ```

Tried that:

```golang
func UnmarshalNew[M Message](data []byte) (M, error) {
	var m M
	err := m.Unmarshal(data)
	return m, err
}

// and then
curVer, err := protoutil.UnmarshalNew[ClusterVersion](encoded.([]byte))
if err != nil {
	log.Fatalf(ctx, "%v", err)
}
```

This does not build:

```
pkg/clusterversion/setting.go:139:40: ClusterVersion does not satisfy protoutil.Message (method MarshalTo has pointer receiver)
```

Setting the `M` type to pointer doesn't help either: `var m M` becomes a nil, and then there is a nil deref panic.

--- a/pkg/clusterversion/setting.go
+++ b/pkg/clusterversion/setting.go
@@ -137,7 +137,7 @@ func (cv *clusterVersionSetting) activeVersionOrEmpty(
 		return ClusterVersion{}
 	}
 	var curVer ClusterVersion
-	if err := protoutil.Unmarshal(encoded.([]byte), &curVer); err != nil {
+	if err := curVer.Unmarshal(encoded.([]byte)); err != nil {
 		log.Fatalf(ctx, "%v", err)
 	}
 	return curVer

--- a/pkg/clusterversion/setting.go
+++ b/pkg/clusterversion/setting.go
@@ -154,7 +154,7 @@ func (cv *clusterVersionSetting) isActive(
 // Decode is part of the VersionSettingImpl interface.
 func (cv *clusterVersionSetting) Decode(val []byte) (settings.ClusterVersionImpl, error) {
 	var clusterVersion ClusterVersion
-	if err := protoutil.Unmarshal(val, &clusterVersion); err != nil {
+	if err := clusterVersion.Unmarshal(val); err != nil {
 		return nil, err
 	}
 	return clusterVersion, nil
@@ -165,7 +165,7 @@ func (cv *clusterVersionSetting) ValidateVersionUpgrade(
 	_ context.Context, sv *settings.Values, curRawProto, newRawProto []byte,
 ) error {
 	var newCV ClusterVersion
-	if err := protoutil.Unmarshal(newRawProto, &newCV); err != nil {
+	if err := newCV.Unmarshal(newRawProto); err != nil {
 		return err
 	}
 
@@ -174,7 +174,7 @@ func (cv *clusterVersionSetting) ValidateVersionUpgrade(
 	}
 
 	var oldCV ClusterVersion
-	if err := protoutil.Unmarshal(curRawProto, &oldCV); err != nil {
+	if err := oldCV.Unmarshal(curRawProto); err != nil {
 		return err
 	}
 
@@ -210,7 +210,7 @@ func (cv *clusterVersionSetting) ValidateBinaryVersions(
 	}()
 
 	var ver ClusterVersion
-	if err := protoutil.Unmarshal(rawProto, &ver); err != nil {
+	if err := ver.Unmarshal(rawProto); err != nil {
 		return err
 	}
 	return cv.validateBinaryVersions(ver.Version, sv)

--- a/pkg/clusterversion/setting_test.go
+++ b/pkg/clusterversion/setting_test.go
@@ -60,3 +60,13 @@ func TestMakeMetricsAndRegisterOnVersionChangeCallback(t *testing.T) {
 		return nil
 	})
 }
+
+func BenchmarkClusterVersionSettingIsActive(b *testing.B) {
+	s := cluster.MakeTestingClusterSettingsWithVersions(clusterversion.TestingBinaryVersion, clusterversion.TestingBinaryMinSupportedVersion, true)
+	ctx := context.Background()
+	active := true
+	for i := 0; i < b.N; i++ {
+		active = s.Version.IsActive(ctx, clusterversion.VCurrent_Start) && active
+	}
+	require.True(b, active)
+}

